safir/src/safir/database/_pagination.py

```diff
@@ -0,0 +1,525 @@
+"""Support for paginated database queries.
+
+This pagination support uses keyset pagination rather than relying on database
+cursors, since the latter interact poorly with horizontally scaled services.
+"""
+
+from __future__ import annotations
+
+from abc import ABCMeta, abstractmethod
+from dataclasses import dataclass
+from datetime import UTC, datetime
+from typing import Generic, Self, TypeVar
+from urllib.parse import parse_qs, urlencode
+
+from pydantic import BaseModel
+from sqlalchemy import Select, and_, func, or_, select
+from sqlalchemy.ext.asyncio import async_scoped_session
+from sqlalchemy.orm import DeclarativeBase, InstrumentedAttribute
+from starlette.datastructures import URL
+
+from ._datetime import datetime_to_db
+
+C = TypeVar("C", bound="PaginationCursor")
+"""Type of a cursor for a paginated list."""
+
+E = TypeVar("E", bound="BaseModel")
+"""Type of an entry in a paginated list."""
+
+__all__ = [
+    "DatetimeIdCursor",
+    "PaginationCursor",
+    "PaginatedList",
+    "PaginatedQueryRunner",
+]
+
+
+@dataclass
+class PaginationCursor(Generic[E], metaclass=ABCMeta):
+    """Generic pagnination cursor for keyset pagination.
+
+    The generic type parameter is the Pydantic model into which each row will
+    be converted, not the ORM model.
+    """
+
+    previous: bool
+    """Whether to search backwards instead of forwards."""
+
+    @classmethod
+    @abstractmethod
+    def from_entry(cls, entry: E, *, reverse: bool = False) -> Self:
+        """Construct a cursor with an entry as a bound.
+
+        Builds a cursor to get the entries after the provided entry, or before
+        the provided entry if ``reverse`` is set to `True`.
+
+        Parameters
+        ----------
+        entry
+            Basis of the cursor.
+        reverse
+            Whether to create a previous cursor.
+
+        Returns
+        -------
+        PaginationCursor
+            Requested cursor.
+        """
+
+    @classmethod
+    @abstractmethod
+    def from_str(cls, cursor: str) -> Self:
+        """Build cursor from the string serialization form.
+
+        Parameters
+        ----------
+        cursor
+            Serialized form of the cursor.
+
+        Returns
+        -------
+        DatetimeIdCursor
```

Comment:
Thanks for catching that! I tend to lean away from using `Self` as the return type in docstrings because the linked documentation is then not very helpful.